flecsi/exec/hpx/policy.hh

```diff
@@ -0,0 +1,231 @@
+// Copyright (c) 2016, Triad National Security, LLC
+// All rights reserved.
+
+#ifndef FLECSI_EXEC_HPX_POLICY_HH
+#define FLECSI_EXEC_HPX_POLICY_HH
+
+#include <flecsi-config.h>
+
+#if !defined(FLECSI_ENABLE_HPX)
+#error FLECSI_ENABLE_HPX not defined! This file depends on HPX!
+#endif
+#if !defined(FLECSI_ENABLE_MPI)
+#error FLECSI_ENABLE_MPI not defined! This file depends on MPI!
+#endif
+
+#include <hpx/modules/collectives.hpp>
+
+#include "flecsi/exec/buffers.hh"
+#include "flecsi/exec/hpx/future.hh"
+#include "flecsi/exec/hpx/reduction_wrapper.hh"
+#include "flecsi/exec/launch.hh"
+#include "flecsi/exec/local/bind_parameters.hh"
+#include "flecsi/exec/prolog.hh"
+#include "flecsi/exec/tracer.hh"
+#include "flecsi/flog.hh"
+#include "flecsi/util/function_traits.hh"
+
+#include <cstddef>
+#include <string>
+#include <type_traits>
+#include <utility> // forward
+
+namespace flecsi {
+namespace exec {
+namespace detail {
+
+template<class T>
+constexpr bool mpi_accessor = false;
+template<data::layout L, class T, Privileges P>
+constexpr bool mpi_accessor<data::accessor<L, T, P>> = !data::portable_v<T>;
+
+// Construct a tuple of converted arguments (all non-references).
+template<bool M, class... PP, class... AA>
+std::conditional_t<M, std::tuple<PP...>, std::tuple<std::decay_t<PP>...>>
+make_parameters(std::tuple<PP...> * p /* to deduce PP */, AA &&... aa) {
```

Comment:
This produces dangling references if an MPI task parameter type `P` is (say) `const float&` and the corresponding argument is `double`.  (Modern Clang/libc++ actually rejects that case, but obviously we want it to work instead.)  The Legion backend had the same bug until 0a75b5dcfacbe27da76682dbc03e5a175a4688cc, which is presumably how it ended up here (sorry).

Reply:
Do we have a test for this case?

Reply:
Do you mean a test that would have undefined behavior without this change?  I don't think so, although it would be trivial to write one.  (Less clear is whether it would demonstrate anything, since undefined behavior is behavior.)

Reply:
I have fixed this and also, I have added a test case for this.